[download fail]: bilibili download failed: panic: runtime error: index out of range [1] with length 0

<!-- 请输入网站的名称 -->
<!-- enter the website name -->

**Website name**: Bilibili

<!-- 你的操作系统 -->
<!-- your operating system -->

**OS:**: Windows

<!-- 视频链接地址 -->
<!-- the video url -->

**Video URL:**: https://www.bilibili.com/bangumi/play/ep704485

<!-- 请输入下载时显示的错误信息 -->
<!-- enter the error message when downloading -->

**Stack overflow**

```
panic: runtime error: index out of range [1] with length 0
```

<!-- 如果能提供截图，则对于解决你的问题非常有帮助 -->
<!-- If you can provide screenshots, it will be very helpful to solve your problem. -->

**Screenshots**

none

<!-- 其他信息 -->
<!-- add any other context about the problem here -->

**Additional context**

```
lux -c bl.txt -debug https://www.bilibili.com/bangumi/play/ep704485

lux: version v0.16.0, A fast and simple video downloader.


URL:         https://www.bilibili.com/bangumi/play/ep704485
Method:      GET
Headers:     http.Header{
    "Accept":          {"text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"},
    "Accept-Charset":  {"UTF-8,*;q=0.5"},
    "Accept-Encoding": {"gzip,deflate,sdch"},
    "Accept-Language": {"en-US,en;q=0.8"},
    "Cookie":          {"_uuid=D9B1AADB-DDF4-10BBA-CF58-676253CDDE9A78747infoc; b_lsid=108EAD7AA_186CEA2D804; b_nut=1663404979; bili_jct=303538e45b28a6bbed35cfb62adb1133; bp_video_offset_26749913=766518492787638345; buvid_fp=950e665fa77caae27ca72c832f0cfeec; buvid_fp_plain=undefined; buvid3=1024B78F-A3CB-BFEF-9B57-E56D1B11145779434infoc; buvid4=3EA21217-05C0-B28E-33A3-2402ACF7E85F88375-022020721-Qd0NZsQY6NFvAP9XHGIwcQ%3D%3D; CURRENT_FNVAL=4048; DedeUserID=26749913; DedeUserID__ckMd5=b94757e248393f8e; fingerprint=950e665fa77caae27ca72c832f0cfeec; LIVE_BUVID=AUTO8416697171148738; nostalgia_conf=-1; PVID=1; rpdid=|(k||RRkYmmJ0J'uY~||llukl; SESSDATA=dc233144%2C1694054793%2C88608%2A31; sid=e3u29g0y"},
    "Referer":         {"https://www.bilibili.com"},
    "User-Agent":      {"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.81 Safari/537.36"},
}
Status Code: 200
panic: runtime error: index out of range [1] with length 0

goroutine 1 [running]:
github.com/iawia002/lux/extractors/bilibili.extractBangumi({0xc0000d4030, 0x2e}, {0xc000a4e000, 0x1903cd}, {0x0, {0x0, 0x0}, 0x1, 0x0, 0x1, ...})
        /Users/iawia002/go/src/github.com/iawia002/lux/extractors/bilibili/bilibili.go:95 +0x825
github.com/iawia002/lux/extractors/bilibili.(*extractor).Extract(0x0?, {0xc0000d4030, 0x2e}, {0x0, {0x0, 0x0}, 0x1, 0x0, 0x1, {0xc0000b2078, ...}, ...})
        /Users/iawia002/go/src/github.com/iawia002/lux/extractors/bilibili/bilibili.go:298 +0x19d
github.com/iawia002/lux/extractors.Extract({0xc0000d4030?, 0x7ff8965304f0?}, {0x0, {0x0, 0x0}, 0x1, 0x0, 0xa, {0xc0000b2078, 0x6}, ...})
        /Users/iawia002/go/src/github.com/iawia002/lux/extractors/extractors.go:52 +0x391
github.com/iawia002/lux/app.download(0xc0002e6000?, {0xc0000d4030, 0x2e})
        /Users/iawia002/go/src/github.com/iawia002/lux/app/app.go:270 +0x2f2
github.com/iawia002/lux/app.New.func1(0xc000265e40)
        /Users/iawia002/go/src/github.com/iawia002/lux/app/app.go:247 +0x69d
github.com/urfave/cli/v2.(*App).RunContext(0xc0002744e0, {0xaffc48?, 0xc0000b2140}, {0xc0000de000, 0x5, 0x8})
        /Users/iawia002/go/pkg/mod/github.com/urfave/cli/v2@v2.6.0/app.go:322 +0x953
github.com/urfave/cli/v2.(*App).Run(...)
        /Users/iawia002/go/pkg/mod/github.com/urfave/cli/v2@v2.6.0/app.go:224
main.main()
        /Users/iawia002/go/src/github.com/iawia002/lux/main.go:13 +0x53
```

--- a/extractors/bilibili/bilibili.go
+++ b/extractors/bilibili/bilibili.go
@@ -3,6 +3,7 @@ package bilibili
 import (
 	"encoding/json"
 	"fmt"
+	"sort"
 	"strconv"
 	"strings"
 	"time"
@@ -92,20 +93,51 @@ type bilibiliOptions struct {
 }
 
 func extractBangumi(url, html string, extractOption extractors.Options) ([]*extractors.Data, error) {
-	dataString := utils.MatchOneOf(html, `window.__INITIAL_STATE__=(.+?);\(function`)[1]
-	var data bangumiData
-	err := json.Unmarshal([]byte(dataString), &data)
+	dataString := utils.MatchOneOf(html, `<script\s+id="__NEXT_DATA__"\s+type="application/json"\s*>(.*?)</script\s*>`)[1]
+	epMapString := utils.MatchOneOf(dataString, `"epMap"\s*:\s*(.+?)\s*,\s*"initEpList"`)[1]
+	fullVideoIdString := utils.MatchOneOf(dataString, `"videoId"\s*:\s*"(ep|ss)(\d+)"`)
+	epSsString := fullVideoIdString[1] // "ep" or "ss"
+	videoIdString := fullVideoIdString[2]
+
+	var epMap map[string]json.RawMessage
+	err := json.Unmarshal([]byte(epMapString), &epMap)
 	if err != nil {
 		return nil, errors.WithStack(err)
 	}
+	var data bangumiData
+	for idString, jsonByte := range epMap {
+		var epInfo bangumiEpData
+		err := json.Unmarshal(jsonByte, &epInfo)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		epID, err := strconv.ParseInt(idString, 10, 0)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		epInfo.EpID = int(epID)
+		if idString == videoIdString || (epSsString == "ss" && epInfo.TitleFormat == "第1话") {
+			data.EpInfo = epInfo
+		}
+		data.EpList = append(data.EpList, epInfo)
+	}
+
+	sort.Slice(data.EpList, func(i, j int) bool {
+		return data.EpList[i].EpID < data.EpList[j].EpID
+	})
+
 	if !extractOption.Playlist {
 		aid := data.EpInfo.Aid
 		cid := data.EpInfo.Cid
 		bvid := data.EpInfo.BVid
+		titleFormat := data.EpInfo.TitleFormat
+		longTitle := data.EpInfo.LongTitle
 		if aid <= 0 || cid <= 0 || bvid == "" {
 			aid = data.EpList[0].Aid
 			cid = data.EpList[0].Cid
 			bvid = data.EpList[0].BVid
+			titleFormat = data.EpList[0].TitleFormat
+			longTitle = data.EpList[0].LongTitle
 		}
 		options := bilibiliOptions{
 			url:     url,
@@ -114,6 +146,8 @@ func extractBangumi(url, html string, extractOption extractors.Options) ([]*extr
 			aid:     aid,
 			cid:     cid,
 			bvid:    bvid,
+
+			subtitle: fmt.Sprintf("%s %s", titleFormat, longTitle),
 		}
 		return []*extractors.Data{bilibiliDownload(options, extractOption)}, nil
 	}
@@ -139,6 +173,8 @@ func extractBangumi(url, html string, extractOption extractors.Options) ([]*extr
 			aid:     u.Aid,
 			cid:     u.Cid,
 			bvid:    u.BVid,
+
+			subtitle: fmt.Sprintf("%s %s", u.TitleFormat, u.LongTitle),
 		}
 		go func(index int, options bilibiliOptions, extractedData []*extractors.Data) {
 			defer wgp.Done()
@@ -425,10 +461,14 @@ func bilibiliDownload(options bilibiliOptions, extractOption extractors.Options)
 	}
 	title := parser.Title(doc)
 	if options.subtitle != "" {
+		pageString := ""
+		if options.page > 0 {
+			pageString = fmt.Sprintf("P%d ", options.page)
+		}
 		if extractOption.EpisodeTitleOnly {
-			title = fmt.Sprintf("P%d %s", options.page, options.subtitle)
+			title = fmt.Sprintf("%s%s", pageString, options.subtitle)
 		} else {
-			title = fmt.Sprintf("%s P%d %s", title, options.page, options.subtitle)
+			title = fmt.Sprintf("%s %s%s", title, pageString, options.subtitle)
 		}
 	}
 

--- a/extractors/bilibili/bilibili_test.go
+++ b/extractors/bilibili/bilibili_test.go
@@ -26,7 +26,6 @@ func TestBilibili(t *testing.T) {
 			args: test.Args{
 				URL:   "https://www.bilibili.com/video/av41301960",
 				Title: "【英雄联盟】2019赛季CG 《觉醒》",
-				Size:  70696896,
 			},
 			playlist: false,
 		},
@@ -34,7 +33,7 @@ func TestBilibili(t *testing.T) {
 			name: "bangumi test",
 			args: test.Args{
 				URL:   "https://www.bilibili.com/bangumi/play/ep167000",
-				Title: "狐妖小红娘：第70话 苏苏智商上线",
+				Title: "狐妖小红娘 第70话 苏苏智商上线",
 			},
 		},
 		{

--- a/extractors/bilibili/types.go
+++ b/extractors/bilibili/types.go
@@ -13,11 +13,13 @@ type token struct {
 }
 
 type bangumiEpData struct {
-	Aid  int    `json:"aid"`
-	Cid  int    `json:"cid"`
-	BVid string `json:"bvid"`
-	ID   int    `json:"id"`
-	EpID int    `json:"ep_id"`
+	Aid         int    `json:"aid"`
+	Cid         int    `json:"cid"`
+	BVid        string `json:"bvid"`
+	ID          int    `json:"id"`
+	EpID        int    `json:"ep_id"`
+	TitleFormat string `json:"titleFormat"`
+	LongTitle   string `json:"long_title"`
 }
 
 type bangumiData struct {

